db/sequences.c

```diff
@@ -0,0 +1,192 @@
+ *  cannot be found, -1 is returned. If the next value is greater than the max
+ *  or less than min value, for ascending or descending sequences respectively,
+ *  and cycle is not enabled, <error>.
+ *
+ *  @param name char * Name of the sequence
+ *  @param val long long * Reference to output location
+ */
+int seq_next_val(tran_type *tran, char *name, long long *val)
+{
+    sequence_t *seq = getsequencebyname(name);
+    sequence_range_t *temp;
+    sequence_range_t *node;
+    int rc = 0;
+    int bdberr = 0;
+
+    if (seq == NULL) {
+        // Failed to find sequence with specified name
+        logmsg(LOGMSG_ERROR, "Sequence %s cannot be found\n", name);
+        return -1;
+    }
+
+    // Get lock for in-memory object
+    pthread_mutex_lock(&seq->seq_lk);
+
+    // Check for remaining values.
+    // seq->next_val is only valid if remaining values > 0
+    if (seq->range_head == NULL) {
+        pthread_mutex_unlock(&seq->seq_lk);
+
+        // TODO: Check if master, if not go to master
+
+        // Create new node range
+        node = malloc(sizeof(sequence_range_t));
+
+        if (node == NULL) {
+            logmsg(LOGMSG_ERROR,
+                   "can't allocate memory for new chunk for sequence \"%s\"\n",
+                   name);
+            goto done;
+        }
+
+        // No remaining values, allocate new chunk into range
+        rc = bdb_llmeta_get_sequence_chunk(
+            tran, name, seq->min_val, seq->max_val, seq->increment, seq->cycle,
+            seq->chunk_size, &seq->flags, seq->start_val, &seq->next_start_val,
+            node, &bdberr);
+
+        if (rc) {
+            logmsg(LOGMSG_ERROR,
+                   "can't retrive new chunk for sequence \"%s\"\n", name);
+            free(node);
+            goto done;
+        }
+
+        // Get lock for in-memory object
+        pthread_mutex_lock(&seq->seq_lk);
+
+        // Insert node into ranges linked list
+        insert_sequence_range(seq, node);
+    }
+
+    // Dispense next_val
+    *val = seq->range_head->current;
+
+    // Calculate next value to dispense
+    // Check for integer overflow
+    if (overflow_ll_add(seq->range_head->current, seq->increment)) {
+        goto next_range;
+    }
+
+    // Apply increment, no overflow can occur
+    seq->range_head->current += seq->increment;
+
+    // Check for out of range conditions
+    if ((seq->increment > 0) &&
+            (seq->range_head->current > seq->range_head->max_val) ||
+        (seq->increment < 0) &&
+            (seq->range_head->current < seq->range_head->min_val)) {
+        goto next_range;
+    }
+
+done:
+    pthread_mutex_unlock(&seq->seq_lk);
+    return rc;
+
+next_range:
+    // Current head range is empty, move to next range
+    temp = seq->range_head;
+    seq->range_head = temp->next;
+
+    if (temp->next) {
+        free(temp);
+    }
+
+    goto done;
+}
+
+int sequences_master_change()

```

Comment:
Not sure about this: a schema-change should absolutely clear pre-allocated ranges, but we should be able to dole-out previously allocated values across a master change.